src/main/java/br/ufpe/cin/mergers/SemistructuredMerge.java

```diff
@@ -107,167 +119,204 @@ public static String merge(File left, File base, File right, MergeContext contex
+		FSTNode mergeLeftBaseRight = superimpose(mergeLeftBase, right, null, context,
+				SuperimpositionStep.LeftBase_Right);
+
 		removeRemainingBaseNodes(mergeLeftBaseRight, context);
-		mergeMatchedContent(mergeLeftBaseRight, context);		
+		mergeMatchedContent(mergeLeftBaseRight, context);
 
 		context.superImposedTree = mergeLeftBaseRight;
-		
+
 		return context;
 	}
 
 	/**
 	 * Superimposes two given ASTs.
 	 * 
-	 * @param nodeA representing the first tree
-	 * @param nodeB representing the second tree
-	 * @param parent node to be superimposed in (can be null)
+	 * @param nodeA                representing the first tree
+	 * @param nodeB                representing the second tree
+	 * @param parent               node to be superimposed in (can be null)
 	 * @param context
 	 * @param isProcessingBaseTree
 	 * @return superimposed tree
 	 */
-	private static FSTNode superimpose(FSTNode nodeA, FSTNode nodeB, FSTNonTerminal parent, MergeContext context, boolean isProcessingBaseTree) {
-		if (nodeA.compatibleWith(nodeB)) {
-			FSTNode composed = nodeA.getShallowClone();
-			composed.index = nodeB.index;
-			composed.setParent(parent);
-
-			if (nodeA instanceof FSTNonTerminal && nodeB instanceof FSTNonTerminal) {
-				FSTNonTerminal nonterminalA = (FSTNonTerminal) nodeA;
-				FSTNonTerminal nonterminalB = (FSTNonTerminal) nodeB;
-				FSTNonTerminal nonterminalComposed = (FSTNonTerminal) composed;
-
-				/*
-				 * nodes from base or right
-				 */
-				for (FSTNode childB : nonterminalB.getChildren()) { 	
-					FSTNode childA = nonterminalA.getCompatibleChild(childB);
-					if (childA == null) { 								// means that a base node was deleted by left, or that a right node was added
-						FSTNode cloneB = childB.getDeepClone();
-						if (childB.index == -1)
-							childB.index = nodeB.index;
-						cloneB.index = childB.index;
-
-						nonterminalComposed.addChild(cloneB);			// cloneB must be removed afterwards if it is a base node
-
-						if (isProcessingBaseTree) {
-							context.deletedBaseNodes.add(cloneB); 		// base nodes deleted by left
-							context.nodesDeletedByLeft.add(cloneB);
-						} else {
-							context.addedRightNodes.add(cloneB); 		// nodes added by right
-						}
-					} else {
-						if (childA.index == -1)
-							childA.index = nodeA.index;
-						if (childB.index == -1)
-							childB.index = nodeB.index;
-						
-						if(!isProcessingBaseTree && context.addedLeftNodes.contains(childA)){ //duplications
-							context.addedRightNodes.add(childB); 		
-						}
-
-						nonterminalComposed.addChild(superimpose(childA, childB, nonterminalComposed, context, isProcessingBaseTree));
-					}
-				}
+	private static FSTNode superimpose(FSTNode nodeA, FSTNode nodeB, FSTNonTerminal parent, MergeContext context,
+			SuperimpositionStep step) {
+		if (!nodeA.compatibleWith(nodeB))
+			return null;
+
+		if (areBothTerminals(nodeA, nodeB, parent)) {
```

Comment:
@jvcoutinho a versão inicial checa só se left e right são terminais, a refatorada checa também se o parent é terminal. Isso pode ter algum efeito colaretal?

Reply:
@guilhermejccavalcanti the method definition has 

&& parent instanceof FSTNonTerminal;

so it seems OK.

but this doesn't seem to be necessary @jvcoutinho